### DISCUSSION TITLE
Organization level cache sharing

### DISCUSSION THREAD

**Discussion by desrosj** [community/community#52173] [Actions]
### Select Topic Area

Product Feedback

### Body

One thing that I wish dependency caching through actions/cache supported was organization level sharing. If multiple repositories need an identical cache, being able to share it with any repository in the same organization could result in needing less storage overall, and a higher cache hit rate.

**Comment by pepitoenpeligro** [comment]
Hi team, any news here?

**Reply by amanaknows** [reply]
Just another identity crisis on our hands

**Comment by ghost** [comment]
The actions/cache by default scopes the cache to the key and branch within the repository that the action runs. This means that caches are not shared across different repositories or across different branches within the same repository unless they match the exact key and branch.

**Reply by mastarija** [reply]
Yeah. We know that, but it would be nice if we could share the cache between the repositories in the organization.

**Comment by amanaknows** [comment]
> Hi team, any news here?
> 
> 

Blockchain inspired pull requests and mergers on a ipfs network operator... simple design and html trickery but won't lead anywhere. Broadcast

**Reply by mastarija** [reply]
You some kind of bot? Because what you wrote makes no sense in this context.

**Comment by kamzil** [comment]
If we could even share cache between different branches in the same *repository*, that would be great.

**Reply by janbernhardt** [reply]
Cache is shared between multiple branches, if the cache is located on your default branch

**Comment by GregYeo** [comment]
Particular use case for me is dependent resource sharing.

Case 1) Output resource sharing between modules
Not as a service, some generated file would be sharing between application.
For me, I have several sharing core modules dependent from applications. They are building every time but the sharing core modules are rarely updated. The computing power for sharing module is so bad eco.

Case 2) Dependent service sharing
Most of frontend application or micro-service is looking for some dependent service built in the house.
So the application should have to connect the existing server or build itself during CI.

I can caching it by ref or branch or commit for the dependent service but i still need to caching it by each repository. And the dependent service already built passed.

It can be achieved by sharing package of github but it would be only valid sharing strategy after staging level but development cycle because of short lifetime of the reference.

You can just achieve it by your own storage. It can be easily setting up if you are running self-hosted runner because you can control LAN traffic. However, It can be very useful if the service provider offers this function like github action secret sharing on organization scope. 

I think it is really good for github's own cost reduction and i guess that it is easy to be implemented because you already have repository's cache system.

**Comment by dreusel** [comment]
One thing you could possibly do to achieve what you need is to do the actual build for all your repos in one dedicated builder repo. That way you could re-use the same cache accross all your builds. You would use the `workflow_dispatch` or `repository_dispatch` trigger with a personal access token set up with appropriate permissions like so:

triggering repo:
```
- name: Trigger other repo with payload
  run: |
    curl -X POST \
      -H "Accept: application/vnd.github+json" \
      -H "Authorization: Bearer ${{ secrets.PERSONAL_ACCESS_TOKEN }}" \
      https://api.github.com/repos/OTHER_OWNER/OTHER_REPO/dispatches \
      -d '{
        "event_type": "remote-start",
        "client_payload": {
          "source_repo": "${{ github.repository }}",
          "source_sha": "${{ github.sha }}",
          "version": "${{ github.event.release.tag_name }}",
          "environment": "production"
        }
      }'
```

executing repo:
```
jobs:
  handle-remote:
    runs-on: ubuntu-latest
    steps:
      - name: Show payload
        run: |
          echo "Source: ${{ github.event.client_payload.source_repo }}"
          echo "Version: ${{ github.event.client_payload.version }}"
          echo "Environment: ${{ github.event.client_payload.environment }}"

```
The drawback is that the feedback / status reporting becomes less natural, as the status of your originally triggered action will no longer reflect wheter your build actually succeeded; it just reflects whether or not it was succesfully able to call the build action. But there are still ways to report the status:
- using the same mechanism the build repo could trigger an action in the source repo. The build repo would include whether or not the build action has succeeded and any desired feedback (such as a link to the actual build action results) and the 'feedback' action in the source repo would immediately fail/succeed appropriately, with the appropriate output
- The building repo could set a commit check status in the source repo:
```
- name: Set commit status back on source repo
  if: always()  # run even on failure, so we can send a 'failure' status
  run: |
    state="success"
    description="Deployment succeeded"
    if [ "${{ job.status }}" != "success" ]; then
      state="failure"
      description="Deployment failed"
    fi

    curl -X POST \
      -H "Accept: application/vnd.github+json" \
      -H "Authorization: Bearer ${{ secrets.CROSS_REPO_PAT }}" \
      https://api.github.com/repos/${{ github.event.client_payload.source_repo }}/statuses/${{ github.event.client_payload.source_sha }} \
      -d "{
        \"state\": \"$state\",
        \"context\": \"downstream/target-repo-deploy\",
        \"description\": \"$description\",
        \"target_url\": \"${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}\"
      }"
```

You could also make your triggering action poll the build action untill it completed and then complete with the same status of the actual build action, but then you would be billed for the minutes twice.
